File: app/golang/internal/payment/infrastructure/web/payment_controller.go
package web

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	paymentApp "github.com/orkhan/ecommerce/internal/payment/application"
	"github.com/orkhan/ecommerce/internal/shared/application/bus"
	"github.com/orkhan/ecommerce/internal/shared/infrastructure/api"
)

type PaymentController struct {
	cmdBus   *bus.Bus
	queryBus *bus.QueryHandlers
}

func NewPaymentController(cmdBus *bus.Bus, queryBus *bus.QueryHandlers) *PaymentController {
	return &PaymentController{cmdBus: cmdBus, queryBus: queryBus}
}

func (c *PaymentController) Process(ctx *gin.Context) {
	var cmd paymentApp.ProcessPaymentCommand
	if err := ctx.ShouldBindJSON(&cmd); err != nil {
		ctx.Error(err)
		return
	}
	res, err := bus.Dispatch[paymentApp.ProcessPaymentCommand, paymentApp.PaymentResult](
		ctx.Request.Context(), c.cmdBus, cmd)
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.SuccessWithMessage(gin.H{"payment_id": res.ID}, "Ödəniş emal edildi"))
}

func (c *PaymentController) Show(ctx *gin.Context) {
	id, err := uuid.Parse(ctx.Param("id"))
	if err != nil {
		ctx.Error(err)
		return
	}
	dto, err := bus.Ask[paymentApp.GetPaymentQuery, paymentApp.PaymentDTO](
		ctx.Request.Context(), c.queryBus, paymentApp.GetPaymentQuery{PaymentID: id})
	if err != nil {
		ctx.Error(err)
		return
	}
	ctx.JSON(http.StatusOK, api.Success(dto))
}

func (c *PaymentController) RegisterRoutes(authed *gin.RouterGroup) {
	authed.POST("/payments/process", c.Process)
	authed.GET("/payments/:id", c.Show)
}

File: app/spring/src/main/java/az/ecommerce/payment/infrastructure/web/PaymentController.java
package az.ecommerce.payment.infrastructure.web;

import az.ecommerce.payment.application.command.processpayment.ProcessPaymentCommand;
import az.ecommerce.payment.application.dto.PaymentDto;
import az.ecommerce.payment.application.query.getpayment.GetPaymentQuery;
import az.ecommerce.shared.application.bus.CommandBus;
import az.ecommerce.shared.application.bus.QueryBus;
import az.ecommerce.shared.infrastructure.api.ApiResponse;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.UUID;

/**
 * Laravel: app/Http/Controllers/PaymentController.php
 *   POST /api/payments/process — auth + throttle:payment
 *   GET  /api/payments/{id}    — auth
 */
@RestController
@RequestMapping("/api/payments")
@PreAuthorize("isAuthenticated()")
public class PaymentController {

    private final CommandBus commandBus;
    private final QueryBus queryBus;

    public PaymentController(CommandBus commandBus, QueryBus queryBus) {
        this.commandBus = commandBus;
        this.queryBus = queryBus;
    }

    @PostMapping("/process")
    public ApiResponse<Map<String, UUID>> process(@RequestBody @Valid ProcessPaymentCommand cmd) {
        UUID id = commandBus.dispatch(cmd);
        return ApiResponse.success(Map.of("payment_id", id), "Ödəniş emal edildi");
    }

    @GetMapping("/{id}")
    public ApiResponse<PaymentDto> show(@PathVariable UUID id) {
        return ApiResponse.success(queryBus.ask(new GetPaymentQuery(id)));
    }
}
